plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/ui/LoginServiceUiTest.java

```diff
@@ -0,0 +1,117 @@
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.Semaphore;
+
+public class LoginServiceUiTest {
+
+  @Test
+  public void testScheduleCodeWaitingJob_successfulLogin()
+      throws IOException, InterruptedException {
+    LocalServerReceiverWrapper codeReceiver = mock(LocalServerReceiverWrapper.class);
+    when(codeReceiver.waitForCode()).thenReturn("some valid authorization code");
+
+    Semaphore wait = new Semaphore(0);
+    String[] codeHolder = new String[1];
+    IOException[] exceptionHolder = new IOException[1];
+
+    new LoginServiceUi(null, null, null)
+        .scheduleCodeWaitingJob(codeReceiver, wait, codeHolder, exceptionHolder)
+        .join();
+
+    verify(codeReceiver, times(1)).stop();
+    Assert.assertEquals("some valid authorization code", codeHolder[0]);
+    Assert.assertNull(exceptionHolder[0]);
+    Assert.assertEquals(1, wait.availablePermits());
+  }
+
+  @Test
+  public void testScheduleCodeWaitingJob_failedLogin()
+      throws IOException, InterruptedException {
+    LocalServerReceiverWrapper codeReceiver = mock(LocalServerReceiverWrapper.class);
+    when(codeReceiver.waitForCode()).thenThrow(new IOException("some IOException"));
+
+    Semaphore wait = new Semaphore(0);
+    String[] codeHolder = new String[1];
+    IOException[] exceptionHolder = new IOException[1];
+
+    new LoginServiceUi(null, null, null)
+        .scheduleCodeWaitingJob(codeReceiver, wait, codeHolder, exceptionHolder)
+        .join();
+
+    verify(codeReceiver, times(1)).stop();
+    Assert.assertNull(codeHolder[0]);
+    Assert.assertEquals("some IOException", exceptionHolder[0].getMessage());
+    Assert.assertEquals(1, wait.availablePermits());
+  }
+
+  private boolean cancelRequestReceived;
+
+  @Test
+  public void testStopCodeWaitingJob() throws IOException, InterruptedException {
+    try (ServerSocket serverSocket = new ServerSocket()) {
+      serverSocket.bind(null);
+      Thread serverThread = createListenerThread(serverSocket);
+      serverThread.start();
+
+      new LoginServiceUi(null, null, null)
+          .stopCodeWaitingJob("http://127.0.0.1:" + serverSocket.getLocalPort());
+
+      serverThread.join(5000);  // Test should pass right away. Don't wait for too long.
+      Assert.assertTrue(cancelRequestReceived);
+    }
+  }
+
+  private Thread createListenerThread(final ServerSocket serverSocket) {
+    return new Thread(new Runnable() {
+      @Override
+      public void run() {
+        try (
+          Socket socket = serverSocket.accept();
+          InputStreamReader reader = new InputStreamReader(socket.getInputStream(), "UTF-8");
+          OutputStreamWriter writer = new OutputStreamWriter(socket.getOutputStream(), "UTF-8");
+        ) {
+          String input = new String();
+          while (true) {
```

Comment:
you need to test for end-of stream (-1) here; otherwise this can block indefinitely if you don't get the expected response